pkgs/servers/http/nginx/nix-etag-1.x.patch

```diff
@@ -0,0 +1,167 @@
+     r->headers_out.content_length_n = len;
+     r->headers_out.last_modified_time = of.mtime;
+ 
+-    if (ngx_http_set_etag(r) != NGX_OK) {
++    if (ngx_http_set_etag_p(r, &path) != NGX_OK) {
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+     }
+ 
+diff --git a/src/http/modules/ngx_http_gzip_static_module.c b/src/http/modules/ngx_http_gzip_static_module.c
+index 7652a9af..b5d1721f 100644
+--- a/src/http/modules/ngx_http_gzip_static_module.c
++++ b/src/http/modules/ngx_http_gzip_static_module.c
+@@ -228,7 +228,7 @@ ngx_http_gzip_static_handler(ngx_http_request_t *r)
+     r->headers_out.content_length_n = of.size;
+     r->headers_out.last_modified_time = of.mtime;
+ 
+-    if (ngx_http_set_etag(r) != NGX_OK) {
++    if (ngx_http_set_etag_p(r, &path) != NGX_OK) {
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+     }
+ 
+diff --git a/src/http/modules/ngx_http_mp4_module.c b/src/http/modules/ngx_http_mp4_module.c
+index 618bf787..6bcaa5b8 100644
+--- a/src/http/modules/ngx_http_mp4_module.c
++++ b/src/http/modules/ngx_http_mp4_module.c
+@@ -634,7 +634,7 @@ ngx_http_mp4_handler(ngx_http_request_t *r)
+     r->headers_out.status = NGX_HTTP_OK;
+     r->headers_out.last_modified_time = of.mtime;
+ 
+-    if (ngx_http_set_etag(r) != NGX_OK) {
++    if (ngx_http_set_etag_p(r, &path) != NGX_OK) {
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+     }
+ 
+diff --git a/src/http/modules/ngx_http_static_module.c b/src/http/modules/ngx_http_static_module.c
+index 0e16c05a..0015f78c 100644
+--- a/src/http/modules/ngx_http_static_module.c
++++ b/src/http/modules/ngx_http_static_module.c
+@@ -218,7 +218,7 @@ ngx_http_static_handler(ngx_http_request_t *r)
+     r->headers_out.content_length_n = of.size;
+     r->headers_out.last_modified_time = of.mtime;
+ 
+-    if (ngx_http_set_etag(r) != NGX_OK) {
++    if (ngx_http_set_etag_p(r, &path) != NGX_OK) {
+         return NGX_HTTP_INTERNAL_SERVER_ERROR;
+     }
+ 
+diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
+index cb49ef74..3ef39380 100644
+--- a/src/http/ngx_http_core_module.c
++++ b/src/http/ngx_http_core_module.c
+@@ -1577,37 +1577,88 @@ ngx_http_set_exten(ngx_http_request_t *r)
+     return;
+ }
+ 
+-
+ ngx_int_t
+ ngx_http_set_etag(ngx_http_request_t *r)
++{
++    return ngx_http_set_etag_p(r, NULL);
++}
++
++ngx_int_t
++ngx_http_set_etag_p(ngx_http_request_t *r, ngx_str_t *path)
+ {
+     ngx_table_elt_t           *etag;
+     ngx_http_core_loc_conf_t  *clcf;
++    u_char                    *real = NULL, *ptr1, *ptr2;
+ 
+     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
+ 
+     if (!clcf->etag) {
+         return NGX_OK;
+     }
++    if (r->headers_out.last_modified_time == 1 && path != NULL) {
++        real = ngx_realpath(path->data, NULL);
++
++        if (real == NULL) {
++            return NGX_OK;
++        }
++    }
+ 
+     etag = ngx_list_push(&r->headers_out.headers);
+     if (etag == NULL) {
++        ngx_free(real);
```

Comment:
If you do `NULL` checks before `free`, this seems reachable without setting `real` when `path` is `NULL`